src/Storages/KeyDescription.cpp

```diff
@@ -167,6 +172,28 @@ KeyDescription KeyDescription::getSortingKeyFromAST(
         result.sample_block = ExpressionAnalyzer(expr, syntax_result, context).getActions(true)->getSampleBlock();
     }
 
+    if (loading_strictness_level <= LoadingStrictnessLevel::CREATE)
+    {
+        /// Forbid IN (subquery). It leads to "Not-ready Set".
+        for (const auto & node : result.expression->getActionsDAG().getNodes())
+        {
+            if (node.column)
+            {
+                const ColumnSet * column_set = checkAndGetColumnConstData<const ColumnSet>(node.column.get());
+                if (!column_set)
+                    column_set = checkAndGetColumn<const ColumnSet>(node.column.get());
+
+                if (!column_set)
+                    continue;
+
+                auto future_set = column_set->getData();
+                if (!future_set || !future_set->get())
+                    throw Exception(ErrorCodes::DATA_TYPE_CANNOT_BE_USED_IN_KEY,
+                        "IN function is not allowed in key expression");
```

Comment:
@KochetovNicolai, this is obviously wrong, because we should allow expressions like `x IN (1, 2, 3), x` in the primary key. Let's correct the message and add a test.